docs/RFCS/20210604_distributed_token_bucket.md

```diff
@@ -0,0 +1,416 @@
+via an internal SQL function; we can easily provide other endpoints as
+necessary.
+
+```sql
+SELECT crdb_internal.update_tenant_resource_limits(
+  tenant_id,                    -- INT
+  operation_uuid,               -- UUID
+  as_of,                        -- TIMESTAMP
+  as_of_consumed_request_units, -- FLOAT
+  available_request_units,      -- FLOAT
+  refill_rate,                  -- FLOAT
+  max_request_units             -- FLOAT
+)
+```
+
+The statement completes after a new row is inserted in the system table.
+
+Updated settings are calculated based on the total consumption reported at some
+recent point in the past. The consumption value is reported as part of the
+reconfiguration request so that we can account for what happened since that
+value was reported; tokens are adjusted based on
+ - the delta between the previous consumption and the current consumption; and
+ - the refill that would have happened in the delta time period.
+
+This API should be initially called as part of creation of a new tenant (with
+`as_of_consumed_request_units = 0`). The system can start with some reasonable
+defaults (e.g. free tier settings) until that happens.
+
+In the initial implementation, this function will be called separately for each
+tenant. We will investigate batching opportunities at a later point.
+
+## Resilience considerations
+
+The system must have reasonable behavior if the bucket range becomes temporarily
+inaccessible. To achieve this, in the short term each node continues operating
+at the previous rate of consumption. Longer term, the rate can decay over time
+toward 1/N of the total refill rate, where N is the number of SQL pods.
+
+## Performance considerations
+
+The system table could get a high amount of aggregate traffic. If we have 5,000
+SQL pods across all tenants, we expect 500 operations per second (for a 10s
+target request period).
+
+Load-based splitting of ranges would reduce the range sizes; we could also force
+split points between the tenants.
+
+The minimum refill amount can be tuned to reduce the frequency of bucket
+operations during periods of low activity on a single node.
+
+## Drawbacks
+
+- Complexity
+
+- Periodic transactions on the host cluster
+
+## Rationale and Alternatives
+
+We have explored a number of existing systems for cloud rate limiting. Most of
+these use a gossip-style algorithm to disseminate relative usage information.
+This is not necessary in our system which provides consistent access to shared
+state.
+
+An alternative to using transactions as the only centralization point would be
+to elect one SQL pod which implements the distributed bucket logic and which all
+other pods communicate with (perhaps even in a streaming fashion). This SQL pod
+would update the persistent state (at a reduced frequency). The difficulty with
+this approach is around implementing election reliably; this could be done by
+keeping track of a "lease" in the bucket state (in the system table).
```

Comment:
Another annoying thing is that if a pod writes to KV for the distributed bucket, these writes need to bypass the bucket.
I feel like the pods will elect a leader for some reason at some point anyway, so I don't really buy a) that this is hard and b) that it wouldn't be worth doing anyway. Not to say that I think it should be done that way, but it's worth looking at the pod leader as something that might be a good idea in general.

Reply:
In this proposal, the "leader" would still use this proposed KV connector API, it would just do it less frequently while allowing more frequent re-adjustment between the SQL pods from the tenant. It's something that we can definitely explore as "step 2" after we get the proposed variant working.

Yeah I would guess there will be other usecases for a leader. I don't think it's hard to do either given that we can bootstrap it on top of a system table (which e.g. stores info about a leader lease).